### PR TITLE
Add/update block height keywords in Clarity

### LINKED ISSUE #3943: `block-height` in Nakamoto
This question arose on Discord:
> What happens to BNS name renewals with the Nakamoto upgrade? Will the dramatic increase in block time mean that BNS renewals will be due much sooner?

To clarify that a bit more, in Nakamoto, the block height will be incrementing on the order of seconds, rather than 10 minutes. This would cause BNS, and potentially other software that uses the `block-height` to have unexpected behavior (e.g. much sooner expiration than expected).

One proposal is to make `block-height` in Clarity 1 or Clarity 2 contracts continue to increment at a similar pace by incrementing on tenure changes, instead of at each new block. In Nakamoto, the tenure changes once per Bitcoin block, so it would be similar to the current block-height pace. In Clarity 3, `block-height`would still indicate the height.

Two problems with this proposal are:
1. It could be confusing for `block-height` to have different meanings in different versions of Clarity
2. Functions that take in a block height, such as `at-block` or `get-block-info?` would not work as expected if passed a tenure height instead of a block height.

### PR DESCRIPTION
**Pull request by obycode**
### Description

Adding support for `stacks-block-height` and `tenure-height` in Clarity 3. This also includes using the `tenure-height` logic for `block-height` logic in Clarity 1 and 2.

### Applicable issues

- fixes #3943

### Checklist

- [x] Test coverage for new or modified code paths
- [x] Changelog is updated
- [x] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [x] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [x] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

**Comment by obycode** [comment]
I'm still checking on a couple of the integration tests, but @kantai, when you have some time, could you take a look at the approach here and let me know if you see anything troublesome. 🙏

**Comment by friedger** [comment]
Somehow related is the get-tenure-info https://github.com/stacks-network/stacks-core/issues/4716
Is that in scope for Clarity 3?

File: clarity/src/vm/contexts.rs [hunk 1]
```diff
@@ -822,6 +822,16 @@ impl<'a, 'hooks> OwnedEnvironment<'a, 'hooks> {
             self.context.eval_hooks = Some(vec![hook]);
         }
     }
+
+    #[cfg(any(test, feature = "testing"))]
```
Comment (jcnelson) [inline clarity/src/vm/contexts.rs:826]: Nit: This should go into the test module

File: clarity/src/vm/database/clarity_db.rs [hunk 2]
```diff
@@ -855,6 +855,31 @@ impl<'a> ClarityDatabase<'a> {
         Ok(())
     }
 
+    fn tenure_height_key() -> &'static str {
+        "_stx-data::tenure_height"
```
Comment (jcnelson) [inline clarity/src/vm/database/clarity_db.rs:859]: Can you make sure SIP-021 reflects this new MARF key?  We've done so for the other MARF keys in the Stacks chainstate MARF.

File: clarity/src/vm/variables.rs [hunk 3]
```diff
@@ -22,31 +24,37 @@ use crate::vm::errors::{InterpreterResult as Result, RuntimeErrorType};
...
     ) -> Option<NativeVariables> {
-        NativeVariables::lookup_by_name(name).and_then(|native_function| {
-            if &native_function.get_version() <= version {
-                Some(native_function)
-            } else {
-                None
+        NativeVariables::lookup_by_name(name).and_then(|native_variable| {
+            match native_variable.get_max_version() {
+                Some(ref max_version)
+                    if &native_variable.get_min_version() <= version && version <= max_version =>
```
Comment (jcnelson) [inline clarity/src/vm/variables.rs:52]: Huh, I didn't know this was legal Rust.  TIL.

File: clarity/src/vm/variables.rs [hunk 4]
```diff
@@ -92,8 +100,18 @@ pub fn lookup_reserved_variable(
...
+                // In epoch 2.x, the `block-height` keyword returns the block height, but
+                // beginning in epoch 3, it returns the tenure height instead in order to
+                // maintain a similar pace to the block  height advancement pre-Nakamoto. This
+                // keyword is removed in Clarity 3 to avoid confusion and replaced with
+                // `stacks-block-height` and `tenure-height`.
+                if env.global_context.epoch_id < StacksEpochId::Epoch30 {
+                    let block_height = env.global_context.database.get_current_block_height();
+                    Ok(Some(Value::UInt(block_height as u128)))
+                } else {
+                    let tenure_height = env.global_context.database.get_tenure_height()?;
```
Comment (jcnelson) [inline clarity/src/vm/variables.rs:112]: I had to read this a couple times before it made sense.  The `block-height` variable isn't going away _per se_; it's just not recognized as a global in Clarity 3.  But, it still has to have semantic meaning in both Clarity 1 and Clarity 2 contracts while in epoch 3.0.  This code is meant to handle this case specifically -- we evaluate `block-height` as the tenure height when in epoch 3.0.  This comment might need some massaging. 

Related, can you still deploy Clarity 1 and 2 contracts in epoch 3.0?

File: clarity/src/vm/variables.rs [hunk 5]
```diff
@@ -123,6 +141,16 @@ pub fn lookup_reserved_variable(
                 let chain_id = env.global_context.chain_id;
                 Ok(Some(Value::UInt(chain_id.into())))
             }
+            NativeVariables::StacksBlockHeight => {
```
Comment (jcnelson) [inline clarity/src/vm/variables.rs:144]: These need to be gated

File: stackslib/src/chainstate/stacks/boot/docs.rs [hunk 6]
```diff
@@ -1,4 +1,5 @@
 use clarity::vm::docs::contracts::{produce_docs_refs, ContractSupportDocs};
+use clarity::vm::ClarityVersion;
```
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/docs.rs:2]: missing copyright comment block

File: clarity/src/vm/variables.rs (hunk 5 above)
Comment (obycode) [reply clarity/src/vm/variables.rs:144]: I believe this is already gated by the `lookup_by_name_at_version` that is the source of the `variable`.

File: stackslib/src/clarity_vm/clarity.rs [hunk 7]
```diff
@@ -294,6 +295,26 @@ impl ClarityInstance {
             )
         };
 
+        // If we're starting a new tenure, and we're in epoch 3.09, then we
```
Comment (jcnelson) [inline stackslib/src/clarity_vm/clarity.rs:298]: Typo -- epoch 3.09?

File: stackslib/src/clarity_vm/clarity.rs [hunk 8]
```diff
@@ -294,6 +295,26 @@ impl ClarityInstance {
...
+        // need to increment the tenure height in the Clarity DB.
+        if new_tenure && epoch.epoch_id >= StacksEpochId::Epoch30 {
```
Comment (jcnelson) [inline stackslib/src/clarity_vm/clarity.rs:300]: This seems brittle to me.  If you're going to introduce a new MARF key/value pair anyway for representing the tenure height, can you instead update it as part of the Nakamoto chain state's `append_block` function?

File: clarity/src/vm/variables.rs (hunk 4 above)
Comment (obycode) [reply clarity/src/vm/variables.rs:112]: Yes, your explanation is correct and is what I was trying to get across with that comment. Does this sound better?

```rust
                // In epoch 2.x, the `block-height` keyword returns the Stacks block height.
                // For Clarity 1 and Clarity 2 contracts executing in epoch 3, `block-height`
                // is equal to the tenure height instead of the Stacks block height. This change
                // is made to maintain a similar pace at which this value increments (e.g. for use
                // as an expiration). In Clarity 3, `block-height` is removed to avoid confusion.
                // It is replaced with two new keywords: `stacks-block-height` and `tenure-height`.
```

> Related, can you still deploy Clarity 1 and 2 contracts in epoch 3.0?
 Yes, you can.

File: stackslib/src/clarity_vm/clarity.rs (hunk 8 above)
Comment (obycode) [reply stackslib/src/clarity_vm/clarity.rs:300]: I need the value while building the block, for use while evaluating the transactions included in the block. If we do it in `append_block`, then I think it will be off by 1.

**Review by jcnelson — COMMENTED** [review]
I think there's some more test coverage you'll want to add regarding the interaction between `block-height`, `stacks-block-height`, and `tenure-height`:

* Can you deploy a Clarity 1 or Clarity 2 contract with bound variable names `stacks-block-height` and `tenure-height` in epoch 3.0?

* Can you call into a Clarity 1 and Clarity 2 contract which have bound variable names `stacks-block-height` and `tenure-height` from a Clarity 3 contract, and have those names behave the right way?

* Can a Clarity 1 and Clarity 2 contract take a Clarity 3 trait implementation, whose code body makes use of `stacks-block-height` and `tenure-height`, and have it execute correctly even though the caller is a Clarity 1 or Clarity 2 context?  Can they do so even if they've been deployed before epoch 3?

* Can a Clarity 1 and Clarity 2 contract perform the above using `at-block` to call into the Clarity 3 contract from a block that's before epoch 3.0?

Basically, I'm trying to make sure that these keywords work correctly if control flow passes back and forth across an epoch boundary via either `at-block` or a trait implementation, or a combination of both.

File: stackslib/src/clarity_vm/clarity.rs (hunk 8 above)
Comment (jcnelson) [reply stackslib/src/clarity_vm/clarity.rs:300]: That won't happen?  You can read an inserted a key/value pair into the MARF before committing it, so if you insert the new tenure height before processing Clarity transactions, you should be fine.

File: clarity/src/vm/contexts.rs (hunk 1 above)
Comment (obycode) [reply clarity/src/vm/contexts.rs]: Addressed in 7a63f899e8945a770642517e96324330597dbe87.

File: clarity/src/vm/variables.rs (hunk 4 above)
Comment (obycode) [reply clarity/src/vm/variables.rs:113]: I updated to the above comment in 7a63f899e8945a770642517e96324330597dbe87.

File: stackslib/src/chainstate/stacks/boot/docs.rs (hunk 6 above)
Comment (obycode) [reply stackslib/src/chainstate/stacks/boot/docs.rs:17]: Added in 7a63f899e8945a770642517e96324330597dbe87.

**Comment by obycode** [comment]
Thanks for the review @jcnelson. I've addressed the simple comments and will work on adding those tests you requested. I also added an integration test since you reviewed which was the main thing that I was not yet comfortable with.

File: stackslib/src/clarity_vm/clarity.rs (hunk 8 above)
Comment (obycode) [reply stackslib/src/clarity_vm/clarity.rs:300]: Ah, okay. I was thinking `append_block` was called after the transactions were processed. I'm a little unclear where exactly I would need to put this and where to get the needed information. I can dig into that tomorrow, but if you've got something in mind, the help would be appreciated.

File: stackslib/src/clarity_vm/clarity.rs (hunk 7 above)
Comment (obycode) [reply stackslib/src/clarity_vm/clarity.rs]: Fixed in 7a63f899e8945a770642517e96324330597dbe87.

File: stacks-common/src/util/macros.rs [hunk 9]
```diff
@@ -88,30 +88,50 @@ macro_rules! define_named_enum {
...
-                    $(
-                        $Name::$Variant => $Version,
-                    )*
+                    $(Self::$Variant => $MinVersion,)*
+                }
+            }
+
+            pub fn get_max_version(&self) -> Option<$VerType> {
+                match self {
+                    $(Self::$Variant => $MaxVersion,)*
```
Comment (kantai) [inline stacks-common/src/util/macros.rs:134]: Can you add rustdocs to the functions e.g., `get_min_version` returns the first Clarity version in which `self` is defined, `get_max_version` returns `Some` for the last Clarity version in which `self` is defined, or `None` if `self` is defined in every Clarity version after `get_min_version()`.

File: stackslib/src/clarity_vm/clarity.rs (hunk 8 above)
Comment (kantai) [reply stackslib/src/clarity_vm/clarity.rs:300]: I think the best strategy is to add this to `setup_block` (which gets invoked in both the miner path and the follower path). Something like this should work:

```diff
diff --git a/stackslib/src/chainstate/nakamoto/mod.rs b/stackslib/src/chainstate/nakamoto/mod.rs
index 2dbcd9f3c..5a9a4b2be 100644
--- a/stackslib/src/chainstate/nakamoto/mod.rs
+++ b/stackslib/src/chainstate/nakamoto/mod.rs
@@ -2551,6 +2551,28 @@ impl NakamotoChainState {
             new_tenure,
         );
 
+        if new_tenure {
+            clarity_tx
+                .connection()
+                .as_free_transaction(|clarity_tx_conn| {
+                    clarity_tx_conn.with_clarity_db(|db| {
+                        let tenure_ht = db.get_tenure_height()?;
+                        db.set_tenure_height(
+                            tenure_ht
+                                .checked_add(1)
+                                .expect("FATAL: tenure height overflowed"),
+                        )?;
+                        Ok(())
+                    })
+                })
+                .map_err(|e| {
+                    error!("Failed to set tenure height during block setup";
+                           "error" => ?e,
+                    );
+                    e
+                })?;
+        }
+
         // now that we have access to the ClarityVM, we can account for reward deductions from
         // PoisonMicroblocks if we have new rewards scheduled
         let matured_rewards_opt = matured_rewards_schedule_opt
```

Then you wouldn't need to pass `new_tenure` through the block begin methods.

**Review by kantai — COMMENTED** [review]
Looks good to me, just some comments here and there, also responded on the thread about where the best place to perform the `get/set` is.

**Comment by obycode** [comment]
Thanks for the review @kantai. Will address these comments shortly.

While adding the tests @jcnelson requested, I realized that I never implemented the analysis error when using `block-height` in Clarity 3. That will be included in the next commit. It's a bit weird, because the rest of the reserved words only hit an error at runtime, but this one will hit the error at analysis. In a future version of Clarity, I'd like to make all of these be detected at analysis time, but not now.

File: clarity/src/vm/functions/mod.rs [hunk 10]
```diff
@@ -201,10 +201,14 @@ impl NativeFunctions {
...
-            } else {
-                None
+            match native_function.get_max_version() {
+                Some(ref max_version)
+                    if &native_function.get_min_version() <= version && version <= max_version =>
+                {
+                    Some(native_function)
+                }
+                None if &native_function.get_min_version() <= version => Some(native_function),
+                _ => None,
```
[the remark this answers is not in the document]
Comment (obycode) [reply clarity/src/vm/functions/mod.rs:211]: Sure, I can see how your version is a little easier to read. Will update.

File: clarity/src/vm/variables.rs [hunk 11]
```diff
@@ -22,31 +24,37 @@ use crate::vm::errors::{InterpreterResult as Result, RuntimeErrorType};
...
-                None
+        NativeVariables::lookup_by_name(name).and_then(|native_variable| {
+            match native_variable.get_max_version() {
+                Some(ref max_version)
+                    if &native_variable.get_min_version() <= version && version <= max_version =>
+                {
+                    Some(native_variable)
+                }
+                None if &native_variable.get_min_version() <= version => Some(native_variable),
+                _ => None,
```
[the remark this answers is not in the document]
Comment (obycode) [reply clarity/src/vm/variables.rs:57]: 🧠

[the remark this answers is not in the document]
Comment (obycode) [reply clarity/src/vm/variables.rs]: Added in aef35a1a57499b0ef3dc111ea0ffd58321eb6875.

File: clarity/src/vm/types/mod.rs [hunk 12]
```diff
@@ -777,10 +777,14 @@ impl BlockInfoProperty {
...
-            } else {
-                None
+            match native_function.get_max_version() {
+                Some(ref max_version)
+                    if &native_function.get_min_version() <= version && version <= max_version =>
+                {
+                    Some(native_function)
+                }
+                None if &native_function.get_min_version() <= version => Some(native_function),
+                _ => None,
```
[the remark this answers is not in the document]
Comment (obycode) [reply clarity/src/vm/types/mod.rs]: Refactored into the macro as suggested below.

File: clarity/src/vm/functions/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (obycode) [reply clarity/src/vm/functions/mod.rs]: Refactored into the macro as suggested below.

File: stacks-common/src/util/macros.rs (hunk 9 above)
Comment (obycode) [reply stacks-common/src/util/macros.rs]: Added in aa832234791c8d902c44618dabbf0f5d48a7546d.

File: stackslib/src/clarity_vm/clarity.rs (hunk 8 above)
Comment (obycode) [reply stackslib/src/clarity_vm/clarity.rs:300]: Thanks for this @kantai. Do you think the initialization in `initialize_epoch_3_0` good to stay there?

Comment (obycode) [reply stackslib/src/clarity_vm/clarity.rs]: Moved in 008f73bf4e5926aa1c8b2c747a06014e37e9e102.

File: clarity/src/vm/contexts.rs (hunk 1 above)
Comment (obycode) [reply clarity/src/vm/contexts.rs]: Turns out moving this into the test module creates some difficulty elsewhere because of our use of `#[cfg(any(test, feature = "testing"))]` in various places. I pulled it out of this impl block and put it in its own block with `#[cfg(any(test, feature = "testing"))]` instead.

**Comment by obycode** [comment]
Ok, I added some more tests in 1bff5b7ce7828248672b74731a8855fa787b06b9. I'm not sure why `test_block_heights_across_versions` is failing yet. I need to resolve that and then add one more test for:

> Can a Clarity 1 and Clarity 2 contract perform the above using at-block to call into the Clarity 3 contract from a block that's before epoch 3.0?

File: stackslib/src/clarity_vm/clarity.rs (hunk 8 above)
Comment (jcnelson) [reply stackslib/src/clarity_vm/clarity.rs]: Better yet, you can just call `db.set_tenure_height(coinbase_height)` since it ought to equal to the height of the current tenure.  No need to re-compute a value that's already given!  (cc @kantai for confirmation)

File: stackslib/src/chainstate/nakamoto/mod.rs [hunk 13]
```diff
@@ -2550,6 +2550,43 @@ impl NakamotoChainState {
...
+                                if cfg!(test) {
+                                    // In test mode, the epoch 3.0 initialization does not always
+                                    // get called to initialize this tenure height so we ignore
+                                    // the error here and return a default value.
+                                    warn!("Failed to get tenure height, defaulting to 0");
+                                    0
+                                } else {
+                                    // If not in test mode, this is definitely an error.
+                                    return Err(e.into());
+                                }
```
Comment (obycode) [inline stackslib/src/chainstate/nakamoto/mod.rs:2570]: I don't love this, but I can't find a better way to handle the test cases that do not properly call the `initialize_epoch_*` functions. Let me know if you have a better solution.

**Comment by obycode** [comment]
> Can a Clarity 1 and Clarity 2 contract perform the above using at-block to call into the Clarity 3 contract from a block that's before epoch 3.0?

Is this actually possible? I don't think you can have a `contract-call?` inside an `at-block` because all contract calls are treated as not read-only and only read-only expressions can be inside an `at-block`.

File: stackslib/src/chainstate/nakamoto/mod.rs (hunk 13 above)
Comment (obycode) [reply stackslib/src/chainstate/nakamoto/mod.rs:2570]: Probably, but, for example, I was having trouble figuring out how to call it in stackslib/src/chainstate/nakamoto/coordinator/tests.rs, `advance_to_nakamoto`.

Comment (kantai) [reply stackslib/src/chainstate/nakamoto/mod.rs:2570]: Ah -- I think this was actually a bug.

The `if new_tenure { .. }` block should be _after_ the `process_epoch_transition` invocation.

File: stackslib/src/clarity_vm/clarity.rs (hunk 8 above)
Comment (kantai) [reply stackslib/src/clarity_vm/clarity.rs]: Yeah -- `coinbase_height` should be the same value.

File: stackslib/src/chainstate/nakamoto/mod.rs (hunk 13 above)
Comment (obycode) [reply stackslib/src/chainstate/nakamoto/mod.rs:2570]: 🤦‍♂️

Comment (obycode) [reply stackslib/src/chainstate/nakamoto/mod.rs]: Fixed and cleaned up the tests.

**Comment by obycode** [comment]
image

@wileyj Can we remove the "CI / Rust Format (pull_request_review)" from the required checks?

**Review by kantai — APPROVED** [review]

**Comment by wileyj** [comment]
> image
> @wileyj Can we remove the "CI / Rust Format (pull_request_review)" from the required checks?

it should happen as soon as this is merged: https://github.com/stacks-network/stacks-core/pull/4765

the status checks are valid for any action that triggers the workflow - it's not so fine-grained that we can disable a status check for a specific action (outside of some semi-complicated conditionals for each workflow job).

**Comment by obycode** [comment]
I pushed one small change that makes the mutants testing a bit happier. The other items it reports I think should be fixed in a more generic way, or else we'll end up with too many `mutants::skip` annotations throughout the code. E.g. it doesn't detect the tests using rstest macro templates as tests, and so it reports missed mutants for all of those.

**Review by kantai — DISMISSED** [review]

File: clarity/src/vm/contexts.rs [hunk 14]
```diff
@@ -1974,6 +1964,18 @@ impl CallStack {
     }
 }
 
+#[cfg(any(test, feature = "testing"))]
+impl<'a, 'hooks> OwnedEnvironment<'a, 'hooks> {
```
Comment (jcnelson) [inline clarity/src/vm/contexts.rs:1968]: Can you put this into the test module that uses it?  Thanks

File: clarity/src/vm/database/clarity_db.rs [hunk 15]
```diff
@@ -855,13 +856,9 @@ impl<'a> ClarityDatabase<'a> {
...
     }
 
-    fn tenure_height_key() -> &'static str {
-        "_stx-data::tenure_height"
-    }
-
     /// Returns the tenure height of the current block.
     pub fn get_tenure_height(&mut self) -> Result<u32> {
-        self.get_data(&Self::tenure_height_key())?
+        self.get_data(TENURE_HEIGHT_KEY)?
```
Comment (jcnelson) [inline clarity/src/vm/database/clarity_db.rs:861]: Should this function and `set_tenure_height` below be gated?

File: clarity/src/vm/tests/variables.rs [hunk 16]
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
...
+            // Do not continue with the test if the evaluation failed.
+            return;
+        }
+    }
+
+    assert_eq!(Ok(expected_success), eval_result);
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_block_height(
```
Comment (jcnelson) [inline clarity/src/vm/tests/variables.rs:324]: Do you also want to test the following?

* `(define-constant block-height ...)`
* `(define-trait block-height ...)`
* tuples with a key named `block-height`
* `(define-fungible-token ...)`
* `(define-non-fungible-token ...)`

File: clarity/src/vm/tests/variables.rs [hunk 17]
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
...
+
+    // function
+    expect_contract_error(
+        version,
+        epoch,
+        &mut tl_env_factory,
+        "function",
+        r#"
+        (define-private (block-height) true)
+        (define-private (test-func) (block-height))
```
Comment (jcnelson) [inline clarity/src/vm/tests/variables.rs:450]: Do you also want to test `(define-public)` and `(define-read-only)`?

File: clarity/src/vm/tests/variables.rs [hunk 18]
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
...
+                |version, _| version >= ClarityVersion::Clarity3,
+                CheckErrors::ReservedWord("block-height".to_string()),
+            ),
+        ],
+        Value::Bool(true),
+    );
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_stacks_block_height(
```
Comment (jcnelson) [inline clarity/src/vm/tests/variables.rs:469]: Same concerns here -- there are other Clarity built-ins that let you create named data that need to be tested.

File: clarity/src/vm/tests/variables.rs [hunk 19]
```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
...
+            WhenError::Initialization,
+            |version, _| version >= ClarityVersion::Clarity3,
+            CheckErrors::NameAlreadyUsed("stacks-block-height".to_string()),
+        )],
+        Value::Bool(true),
+    );
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_tenure_height(
```
Comment (jcnelson) [inline clarity/src/vm/tests/variables.rs:579]: Also tagging this for testing with all Clarity built-ins for creating named data.

File: clarity/src/vm/database/clarity_db.rs (hunk 15 above)
Comment (jcnelson) [reply clarity/src/vm/database/clarity_db.rs:861]: Also, should this fall back to getting the block height if the tenure height key is not mapped in this MARF?  Thinking about what could happen if this function is reachable within a `(at-block SOME_BLOCK_BEFORE_NAKAMOTO ..)` closure.

File: clarity/src/vm/variables.rs (hunk 4 above)
Comment (jcnelson) [reply clarity/src/vm/variables.rs:94]: Thanks for the updated explanation.  But, I'm having a hard time convincing myself that the call to `.get_tenure_height()` can _never_ return an `InterpreterError::Expects(..)` due to the tenure height key not being mapped in the MARF tip of the calling context.  For example, what happens if I do the following?

```clarity
;; Can I call this in Nakamoto?
(at-block SOME_BLOCK_BEFORE_NAKAMOTO tenure-height)`
```

What happens with traits defined in Nakamoto but that get called in a pre-Nakamoto context?

```clarity
;; Suppose .whoops-trait is defined in the Nakamoto epoch, and suppose it has this function
;; definition:
;; (define-public (get-tenure-height) (response uint uint))
(use-trait .whoops-trait)

;; Suppose `tenure-getter`'s `get-tenure-height` function simply returns `tenure-height`.
;; How would this work?
(define-read-only (whoops (tenure-getter <whoops-trait>))
   (at-block SOME_BLOCK_BEFORE_NAKAMOTO (contract-call? tenure-getter get-tenure-height))
```

Comment (jcnelson) [reply clarity/src/vm/variables.rs:94]: I have similar questions about `block-height`.  Can these work?

```clarity
;; Can I access `block-height` within an `at-block` closure within Nakamoto as long as the block
;; predates Nakamoto, or is this now forbidden?
(at-block SOME_BLOCK_BEFORE_NAKAMOTO block-height)
```

Again, what about calls to a trait implementation?

```clarity
;; Suppose .whoops-trait is defined in epoch 2.x, and suppose it has this function
;; definition:
;; (define-public (get-block-height) (response uint uint))
(use-trait .whoops-trait)

;; Suppose `block-height-getter`'s `get-block-height` function simply returns `block-height`.
;; How would this work when called in Nakamoto?
(define-public (whoops (block-height-getter <whoops-trait>))
   (at-block SOME_BLOCK_BEFORE_NAKAMOTO (contract-call? block-height-getter get-block-height)))
```

File: stackslib/src/chainstate/nakamoto/mod.rs [hunk 20]
```diff
@@ -2548,9 +2548,29 @@ impl NakamotoChainState {
             &parent_header_hash,
             &MINER_BLOCK_CONSENSUS_HASH,
             &MINER_BLOCK_HEADER_HASH,
-            new_tenure,
         );
 
+        if new_tenure {
```
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/mod.rs:2553]: Do you have tests to verify that the tenure height gets initialized to `(+ u1 block-height)` for the first Nakamoto block?

File: stackslib/src/clarity_vm/tests/contracts.rs [hunk 21]
```diff
@@ -1114,3 +1114,284 @@ fn test_block_heights() {
...
+    let contract_id_e2c2 = QualifiedContractIdentifier::local("epoch-2-clarity-2").unwrap();
+    let contract_id_e3c3 = QualifiedContractIdentifier::local("epoch-3-clarity-3").unwrap();
+
+    let contract_e2c1_2 = r#"
+        (define-trait getter ((get-int () (response uint uint))))
+        (define-public (get-it (get-trait <getter>))
+            (contract-call? get-trait get-int)
+        )
+        "#;
+    let contract_e3c3 = format!(
```
Comment (jcnelson) [inline stackslib/src/clarity_vm/tests/contracts.rs:1274]: In a similar vein, what happens if the trait implementation was deployed in epoch 2.x, and had a bound name `stacks-block-height` (or `tenure-height`), and was called in epoch 3.x?  The VM should continue to treat the epoch 2.x's bound names as if they were variables, and not reserved words, but I just want to see a test that confirms this (it may exist; I just may not have gotten to it yet :))

File: clarity/src/vm/contexts.rs (hunk 14 above)
Comment (obycode) [reply clarity/src/vm/contexts.rs:1968]: I'm not sure who to do that. If I move it into clarity/src/vm/tests/mod.rs, then I cannot access the private field, `self.context`.

File: clarity/src/vm/database/clarity_db.rs (hunk 15 above)
Comment (obycode) [reply clarity/src/vm/database/clarity_db.rs:861]: Ah, ok, I see what you mean. I was thinking that returning that error would be reasonable if you did something like:

```clarity
(at-block (unwrap-panic (get-block-info? id-header-hash u0)) tenure-height)
```

But maybe your idea of returning the block height would be better.

File: clarity/src/vm/contexts.rs (hunk 14 above)
Comment (jcnelson) [reply clarity/src/vm/contexts.rs:1968]: You can set `self.contract` to have `pub(crate)` visibility.

File: clarity/src/vm/database/clarity_db.rs (hunk 15 above)
Comment (jcnelson) [reply clarity/src/vm/database/clarity_db.rs:861]: Ah, I wasn't trying to be prescriptive here.  I was mainly interested in (1) determining if this code could lead to an `InterpreterError::Expects(..)` (or worse), and (2) if so, starting a conversation on what the right behavior ought to be (so we can add regression tests for it).

Comment (jcnelson) [reply clarity/src/vm/database/clarity_db.rs:861]: Also, per your comment earlier, you _can_ have contract-calls within `at-block` as long as they're read-only.  So, it could be possible to call a Clarity 1 or 2 contract from a Clarity 3 function, and it would be possible to do this for either a Nakamoto block or a Stacks epoch2.x block.  I think we should determine what happens if these Clarity 1/2 contracts use `stacks-block-height`, `tenure-height`, and `block-height` when called this way (recalling that Clarity 1/2 contracts can bind these first two global variables as local names, which would now be forbidden).

Comment (obycode) [reply clarity/src/vm/database/clarity_db.rs:861]: Got it. I think it does makes sense to return the block height in that situation. I've got a corresponding test ready for that. I'm a little unsure about whether the gating is necessary for `set_tenure_height`, but I can add it.

File: clarity/src/vm/tests/variables.rs (hunk 16 above)
Comment (obycode) [reply clarity/src/vm/tests/variables.rs:324]: Sure, I was just trying to cover each category, but I can instead cover all possibilities.

Comment (obycode) [reply clarity/src/vm/tests/variables.rs:324]: Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

File: clarity/src/vm/tests/variables.rs (hunk 17 above)
Comment (obycode) [reply clarity/src/vm/tests/variables.rs:450]: Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

File: clarity/src/vm/tests/variables.rs (hunk 18 above)
Comment (obycode) [reply clarity/src/vm/tests/variables.rs:634]: Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

File: clarity/src/vm/tests/variables.rs (hunk 19 above)
Comment (obycode) [reply clarity/src/vm/tests/variables.rs:867]: Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.

File: stackslib/src/chainstate/nakamoto/mod.rs (hunk 20 above)
Comment (obycode) [reply stackslib/src/chainstate/nakamoto/mod.rs]: Are you thinking an integration test for this or some kind of unit test?

File: stackslib/src/clarity_vm/tests/contracts.rs (hunk 21 above)
Comment (obycode) [reply stackslib/src/clarity_vm/tests/contracts.rs:1274]: Added `test_block_heights_across_versions_traits_2_from_3`

File: stackslib/src/chainstate/nakamoto/mod.rs (hunk 20 above)
Comment (obycode) [reply stackslib/src/chainstate/nakamoto/mod.rs]: Added to my integration test in 4fb8bdd86cbe108e2f0406ee52cc1d4879b2eb2e.

**Review by kantai — APPROVED** [review]

**Review by jcnelson — APPROVED** [review]
Is there a test document somewhere that we can update so we can remember to put this PR through its paces on the Nakamoto testnet?

**Comment by obycode** [comment]
> Is there a test document somewhere that we can update so we can remember to put this PR through its paces on the Nakamoto testnet?

We probably should start a checklist like https://github.com/stacks-network/stacks-core/discussions/4620. cc @saralab
